src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DoNotCallEnumerableCastOrOfTypeWithIncompatibleTypesAnalyzerTests.cs

```diff
@@ -0,0 +1,722 @@
+        _ = (new ITree[0]).Cast<Shoe>();
+
+        _ = (new IGrass[0]).Cast<Plant>();
+        _ = (new IGrass[0]).Cast<Tree>();
+        _ = (new IGrass[0]).Cast<Grass>();
+        _ = (new IGrass[0]).Cast<Shoe>();
+
+        // class to interface
+        _ = (new Plant[0]).Cast<ICar>();
+        _ = (new Tree[0]).Cast<ICar>();
+        _ = {|#52:(new Grass[0]).Cast<ICar>()|}; // grass doesn't implement ICar, and is sealed so a subclass couldn't either
+        _ = (new Shoe[0]).Cast<ICar>();
+
+        _ = (new Plant[0]).Cast<IPlant>();
+        _ = (new Tree[0]).Cast<IPlant>();
+        _ = (new Grass[0]).Cast<IPlant>();
+        _ = (new Shoe[0]).Cast<IPlant>();
+
+        _ = (new Plant[0]).Cast<ITree>();
+        _ = (new Tree[0]).Cast<ITree>();
+        _ = {|#53:(new Grass[0]).Cast<ITree>()|};
+        _ = (new Shoe[0]).Cast<ITree>();
+
+        _ = (new Plant[0]).Cast<IGrass>();
+        _ = (new Tree[0]).Cast<IGrass>();
+        _ = (new Grass[0]).Cast<IGrass>();
+        _ = (new Shoe[0]).Cast<IGrass>();
+    }
+}
+",
+
+VerifyCS.Diagnostic(ofTypeRule).WithLocation(10).WithArguments("int", "string"),
+VerifyCS.Diagnostic(castRule).WithLocation(11).WithArguments("string", "int"),
+VerifyCS.Diagnostic(castRule).WithLocation(12).WithArguments("int", "string"),
+
+VerifyCS.Diagnostic(castRule).WithLocation(40).WithArguments("Fruit", "Shoe"),
+VerifyCS.Diagnostic(castRule).WithLocation(41).WithArguments("Apple", "Shoe"),
+VerifyCS.Diagnostic(castRule).WithLocation(42).WithArguments("Shoe", "Fruit"),
+VerifyCS.Diagnostic(castRule).WithLocation(43).WithArguments("Shoe", "Apple"),
+
+VerifyCS.Diagnostic(castRule).WithLocation(50).WithArguments("ICar", "Grass"),
+VerifyCS.Diagnostic(castRule).WithLocation(51).WithArguments("ITree", "Grass"),
+VerifyCS.Diagnostic(castRule).WithLocation(52).WithArguments("Grass", "ICar"),
+VerifyCS.Diagnostic(castRule).WithLocation(53).WithArguments("Grass", "ITree")
+);
+        }
+
+        [Fact]
+        public async Task ArrayCSharp()
+        {
+            Assert.Throws<InvalidCastException>(()
+                => (new int[][] { new int[] { 1 } }).Cast<object[]>().ToArray());
+            Assert.Throws<InvalidCastException>(()
+                => (new[] { 1 }).Cast<string>().ToArray());
+
+            Assert.Throws<InvalidCastException>(()
+                => (new object[][] { Array.Empty<object>() }).Cast<int[]>().ToArray());
+
+            Assert.Throws<InvalidCastException>(()
+                => (new ValueType[][] { Array.Empty<ValueType>() }).Cast<int[]>().ToArray());
+
+            Assert.Throws<InvalidCastException>(()
+                => (new object[][] { Array.Empty<ValueType>() }).Cast<int[]>().ToArray());
+
+            Assert.Throws<InvalidCastException>(()
+                => (new ValueType[][] { Array.Empty<ValueType>() }).Cast<int[]>().ToArray());
+
+            Assert.Throws<InvalidCastException>(()
+                => (new int[][] { Array.Empty<int>() }).Cast<ValueType[]>().ToArray());
```

Comment:
the idea here was two-fold:
1. it was useful to double check the logic while writing the analyzer
2. this could potentially detect if the behavior changes in future runtime/bcl updates. This is pretty unlikely though, so I wouldn't object to removing them.

Reply:
I am OK with keeping, defer to @mavasani